### PR TITLE
Clarify business logic for CallExpr::isResolved()

### PR DESCRIPTION
**Pull request by noakesmichael**
CallExpr:isResolved() currently returns an FnSymbol in one specific situation and NULL in all
other cases.  In practice this function is used as a predicate, a simple accessor to the FnSymbol,
and to obtain the FnSymbol while executing the business logic associated with resolution.  In the
latter case it is not immediately obvious what cases are left to handle when it returns NULL.

This PR

a) splits isResolved in to several different functions that are better aligned with their intended
purpose

b) takes a first step towards converting isResolved to a simple predicate

c) clarifies/verifies the cases in which isResolved() is currently returning NULL during resolution.

There is no change in behavior for the existing compiler but this version would halt() for malformed
AST in cases where the previous implementation would return NULL.

Tested on linux64 and gasnet.

### CODE REVIEW

File: compiler/AST/astutil.cpp [hunk 1]
```diff
@@ -252,16 +251,23 @@ void compute_call_sites() {
     else
       fn->calledBy = new Vec<CallExpr*>();
   }
+
   forv_Vec(CallExpr, call, gCallExprs) {
-    if (FnSymbol* fn = call->isResolved()) {
+    if (call->isEmpty() == true) {
+
+    } else if (FnSymbol* fn = call->isResolved()) {
```
Comment (lydia-duncan) [inline compiler/AST/astutil.cpp]: Do you want this use of isResolved to also turn into a call to theFnSymbol?

Comment (noakesmichael) [inline compiler/AST/astutil.cpp]: That's a good question; I'm not certain how that's going to turn out.

Right now compute_call_sites() is used during resolution and optimization.  I had to
add the isEmpty() to protect isResolved(), AKA resolvedFunction(), from empty call exprs.
I think for now I am "ok" with this being resolvedFunction() but that might change as I
understand things better.

File: compiler/AST/expr.cpp [hunk 2]
```diff
@@ -3722,9 +3722,67 @@ CallExpr::insertAtTail(BaseAST* ast) {
...
+    if (isUnresolvedSymExpr(baseExpr) == true) {
+
+    } else if (SymExpr* base = toSymExpr(baseExpr)) {
+      if (FnSymbol* fn = toFnSymbol(base->var)) {
+        retval = fn;
+
+      // Probably an array index
+      } else if (isArgSymbol(base->var)  == true ||
+                 isVarSymbol(base->var)  == true) {
+
```
Comment (lydia-duncan) [inline compiler/AST/expr.cpp]: I'm not a big fan of empty else branches, but I see what you're trying to accomplish here

File: compiler/AST/expr.cpp [hunk 3]
```diff
@@ -3722,9 +3722,67 @@ CallExpr::insertAtTail(BaseAST* ast) {
...
-FnSymbol* CallExpr::isResolved(void) {
-  SymExpr* base = toSymExpr(baseExpr);
-  return base ? toFnSymbol(base->var) : NULL;
+bool CallExpr::isEmpty() const {
+  return primitive == NULL && baseExpr == NULL;
+}
+
+
+// MDN 2016/01/29: This will become a predicate
+FnSymbol* CallExpr::isResolved() const {
```
Comment (lydia-duncan) [inline compiler/AST/expr.cpp]: To me, this function should really return a boolean and calls to isResolved() that want the function it resolves to should just call resolvedFunction() instead.  But that'd likely touch a larger swathe of code so I understand why you didn't.  Still seems unfortunate

Comment (noakesmichael) [inline compiler/AST/expr.cpp]: Right; that's the next series of steps.

1) Go through the code base converting uses of isResolved() to resolvedFunction() where appropriate and in
manageable pieces

2) Convert isResolved() to return a bool

File: compiler/AST/expr.cpp [hunk 4]
```diff
@@ -3722,9 +3722,67 @@ CallExpr::insertAtTail(BaseAST* ast) {
...
+  return resolvedFunction();
+}
+
+
+FnSymbol* CallExpr::resolvedFunction() const {
```
Comment (lydia-duncan) [inline compiler/AST/expr.cpp]: It seems like this method is doing some checks that might be occurring in other places.  Are there places where it would be reasonable to remove some similar code to prevent repeated work?

File: compiler/AST/expr.cpp (hunk 3 above)
Comment (lydia-duncan) [inline compiler/AST/expr.cpp]: Ah okay.  I take it you'd like to do that work in a separate PR?  This work is good to go in that case.

File: compiler/AST/expr.cpp (hunk 4 above)
Comment (noakesmichael) [inline compiler/AST/expr.cpp]: You mean resolvedFunction()?

It is true that resolvedFunction has been written in a manner that might do "some repeated work" esp relative
to all of the current calls to isResolved(). However I think it is unlikely that the impact of any such work is
measurable in the current compiler.

In the immediate term it helps to clarify these cases in which isResolved() is returning NULL.
In the middle term it helps to provide some cheap validation as existing uses of isResolved() are being revised.
In the long term the function might be streamlined a bit.

File: compiler/AST/expr.cpp (hunk 3 above)
Comment (noakesmichael) [inline compiler/AST/expr.cpp]: Likely in multiple PRs

File: compiler/AST/expr.cpp (hunk 4 above)
Comment (lydia-duncan) [inline compiler/AST/expr.cpp]: I'm not worried about algorithmic cost of the repeated work, more maintenance cost and the view that going forward it seems like you want that logic in resolvedFunction() rather than outside it.  Removing any potential duplication would make it more likely that additional checks would be inserted inside the method (but it sounds like you already want to do some clean up so I guess I'm agreeing with you?)
